autoload/go/def.vim

```diff
@@ -71,41 +80,191 @@ function! s:godefJump(out, mode)
 	let old_errorformat = &errorformat
 	let &errorformat = "%f:%l:%c"
 
-	if a:out =~ 'godef: '
-		let out = substitute(a:out, go#util#LineEnding() . '$', '', '')
-		echom out
+	" Location is the first line of godef output. Ideally in the proper format
+	" but it could also be an error
+	let location = a:out[0]
+
+	" Echo the godef error if we had one.
+	if location =~ 'godef: '
+		let gderr=substitute(location, go#util#LineEnding() . '$', '', '')
+        call go#util#EchoError(gderr)
+        return
+    endif
+
+    let parts = split(a:out[0], ':')
+
+    " parts[0] contains filename
+    let fileName = parts[0]
+
+    " Don't jump if it's the same identifier we just jumped to
+    if len(w:go_stack) > 0 && w:go_stack[w:go_stack_level-1]['ident'] == a:out[1] && w:go_stack[w:go_stack_level-1]['file'] == fileName
+        return
+    endif
+
+    " needed for restoring back user setting this is because there are two
+    " modes of switchbuf which we need based on the split mode
+    let old_switchbuf = &switchbuf
+
+    " If we got a mode, then just open the def in a new tab/split
+    if a:mode == "tab"
+        let &switchbuf = "usetab"
+
+        if bufloaded(fileName) == 0
+            tab split
+        endif
+    elseif a:mode  == "split"
+        split
+    elseif a:mode == "vsplit"
+        vsplit
+    else
+        " We didn't get a mode, so we start the process of pushing to the
+        " jumpstack
```

Comment:
That's correct. I was trying to avoid inconsistent behavior, but I have just thought of a better way to do this that will push the location AFTER the jump, so that each new window gets its own stack that starts from the first location visited.